posthog/models/cohort.py

```diff
@@ -102,39 +110,39 @@ def get_analytics_metadata(self):
             "deleted": self.deleted,
         }
 
-    def calculate_people(self):
+    def calculate_people(self, batch_size=50000, pg_batch_size=1000):
         if self.is_static:
             return
         try:
-            persons_query = self._clickhouse_persons_query()
-
-            try:
-                sql, params = persons_query.distinct("pk").only("pk").query.sql_with_params()
-            except EmptyResultSet:
-                query = DELETE_QUERY.format(cohort_id=self.pk)
-                params = {}
-            else:
-                query = f"""
-                    {DELETE_QUERY};
-                    {UPDATE_QUERY};
-                """.format(
-                    cohort_id=self.pk,
-                    values_query=sql.replace('FROM "posthog_person"', f', {self.pk} FROM "posthog_person"', 1,),
-                )
-
-            cursor = connection.cursor()
+            start_time = time.time()
+            # Get or wait to update cohort version
             with transaction.atomic():
-                cursor.execute(query, params)
+                cohort = self.queryset().select_for_update().get()
+                cohort.version = cohort.version + 1
+                cohort.save(update_fields=["version"])
+                new_version = cohort.version
+
+            # Paginate fetch batch_size from clickhouse and paginate insert pg_batch_size into postgres
+            cursor = 0
+            persons = self._clickhouse_persons_query(batch_size=batch_size, offset=cursor)
+            while persons:
+                to_insert = [CohortPeople(person_id=p.pk, cohort_id=self.pk, version=new_version) for p in persons]
+                CohortPeople.objects.bulk_create(to_insert, batch_size=pg_batch_size)
+
+                cursor += batch_size
+                persons = self._clickhouse_persons_query(batch_size=batch_size, offset=cursor)
+
+            logger.info("Calculating cohort {} took {:.2f} seconds".format(cohort.pk, (time.time() - start_time)))
```

Comment:
Nit: to use structured logging you could also do something like:

```python
logger.info("cohort_calculation_completed", id=cohort.pk, duration=time.time() - start_time)
```

see [docs](https://posthog.com/docs/contribute/coding-conventions)

Reply:
It's probably worth mentioning that if we do this then we can, for instance, machine read this data and plot e.g. average cohort calculation duration over time, or by cohort pk, or by org if we add that information in.